packages/providers/src/provider.ts

```diff
@@ -280,12 +291,35 @@ export default class Provider {
     };
   }
 
+  addMissingVariableOutputs = async (
+    transactionRequest: TransactionRequest,
+    tries: number = 0
+  ): Promise<void> => {
+    let missingOutputVariableCount = 0;
+
+    if (transactionRequest.type === TransactionType.Create) {
+      return;
+    }
+
+    do {
+      const encodedTransaction = hexlify(transactionRequest.toTransactionBytes());
+      const { dryRun: gqlReceipts } = await this.operations.dryRun({
+        encodedTransaction,
+        utxoValidation: false,
+      });
+      const receipts = gqlReceipts.map(processGqlReceipt);
+      missingOutputVariableCount = getReceiptsWithMissingOutputVariables(receipts).length;
+      transactionRequest.addVariableOutputs(missingOutputVariableCount);
+    } while (tries > MAX_RETRIES || missingOutputVariableCount > 0);
+  };
+
   /**
    * Executes a signed transaction without applying the states changes
    * on the chain.
    */
   async simulate(transactionRequestLike: TransactionRequestLike): Promise<CallResult> {
     const transactionRequest = transactionRequestify(transactionRequestLike);
+    await this.addMissingVariableOutputs(transactionRequest);
```

Comment:
Since `simulate` and `dryRun` are both mutating its underlying transaction, we should make a note of that in the function-level comments; Rust solves this by having the `mut tx` argument, which implies it's going to mutate. Since TS doesn't have the same concept, we should at least make it explicit in the comment!

Reply:
done!